internal/lib/stats/metrics.go

```diff
@@ -0,0 +1,445 @@
+package statsserver
+
+import (
+	"reflect"
+	"time"
+
+	"github.com/cri-o/cri-o/internal/lib/sandbox"
+	"github.com/cri-o/cri-o/internal/oci"
+	types "k8s.io/cri-api/pkg/apis/runtime/v1"
+)
+
+var baseLabelKeys = []string{"id", "name", "image"}
+
+const NetworkMetrics = "network"
+
+type metricValue struct {
+	value      uint64
+	labels     []string
+	metricType types.MetricType
+}
+
+type metricValues []metricValue
+
+type ContainerStats struct {
+	desc      *types.MetricDescriptor
+	valueFunc func() metricValues
+}
+
+type SandboxMetrics struct {
+	metric *types.PodSandboxMetrics
+}
+
+func (s *SandboxMetrics) GetMetric() *types.PodSandboxMetrics {
+	return s.metric
+}
+
+func NewSandboxMetrics(sb *sandbox.Sandbox) *SandboxMetrics {
+	return &SandboxMetrics{
+		metric: &types.PodSandboxMetrics{
+			PodSandboxId:     sb.ID(),
+			Metrics:          []*types.Metric{},
+			ContainerMetrics: []*types.ContainerMetrics{},
+		},
+	}
+}
+
+func (s *SandboxMetrics) ResetContainerMetricsForSandbox() {
+	for _, cm := range s.metric.ContainerMetrics {
+		cm.Metrics = []*types.Metric{} // Reset metrics for each container
+	}
+	s.metric.Metrics = []*types.Metric{} // Reset metrics for the next iteration
+}
+
+func (s *SandboxMetrics) ResetMetricsForSandbox() {
+	s.metric.Metrics = []*types.Metric{} // Reset metrics for the next iteration
+}
+
+// AddMetricToSandboxMetrics adds the metrics for the specified pod/container(s).
+func (s *SandboxMetrics) AddMetricToSandboxMetrics(containerID string, m []*types.Metric) {
+	if containerID == "" {
+		s.metric.Metrics = append(s.metric.Metrics, m...)
+	} else {
+		if cMetrics := findExistingContainerMetric(s.metric.ContainerMetrics, containerID); cMetrics != nil {
+			cMetrics.Metrics = append(cMetrics.Metrics, m...)
+		}
+	}
+}
+
+func findExistingContainerMetric(containerMetrics []*types.ContainerMetrics, containerID string) *types.ContainerMetrics {
+	for _, cm := range containerMetrics {
+		if cm.ContainerId == containerID {
```

Comment:
hm I'm not sure about looping through the container metrics to save on cpu time of recreating the structure. It may be better to keep it simple to start and iterate later if needed